src/resources.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Structs to manage device resources.
+//!
+//! The high level flow of resource management among the VMM, the device manager, and the device
+//! is as below:
+//! 1) the VMM creates a new device object.
+//! 2) the VMM asks the new device object for its resource constraints.
+//! 3) the VMM allocates resources for the device object according to resource constraints.
+//! 4) the VMM passes the allocated resources to the device object.
+//! 5) the VMM registers the new device onto corresponding device managers according the allocated
+//!    resources.
+
+use std::{u16, u32, u64};
+
+/// Enumeration describing a device's resource constraints.
+pub enum ResourceConstraint {
+    /// Constraint for an IO Port address range.
+    PioAddress {
+        /// Allocate resource within the range [`min`, `max`] if `min` is not u16::MAX.
+        min: u16,
+        /// Allocate resource within the range [`min`, `max`] if `max` is not zero.
+        max: u16,
+        /// Alignment for the allocated address.
+        align: u16,
+        /// Size for the allocated address range.
+        size: u16,
+    },
+    /// Constraint for a Memory Mapped IO address range.
+    MmioAddress {
+        /// Allocate resource within the range [`min`, `max`] if 'min' is not u64::MAX.
+        min: u64,
+        /// Allocate resource within the range [`min`, `max`] if 'max' is not zero.
+        max: u64,
+        /// Alignment for the allocated address.
+        align: u64,
+        /// Size for the allocated address range.
+        size: u64,
+    },
+    /// Constraint for a legacy IRQ.
+    LegacyIrq {
+        /// Reserve the preallocated IRQ if it's not u32::MAX.
+        fixed: u32,
```

Comment:
'fixed' does not sounds a good name to me. probably just 'irq'?

Reply:
Fixed.